torch/distributed/algorithms/join.py

```diff
@@ -0,0 +1,210 @@
+        pass.
+
+    .. warning::
+        If ``throw_on_early_termination`` is enabled, then the context manager
+        additionally requires every non-joined process to participate in an
+        all-reduce before it performs its collective communications in order to
+        check if it should terminate due to detecting uneven inputs. This all-
+        reduce should be of the form ``dist.all_reduce(torch.zeros(1))``; if
+        the result is positive, then the process should terminate.
+
+    .. warning::
+        The context manager requires that all ``process_group`` attributes in
+        the ``_JoinHook`` objects are the same. If there are multiple
+        ``_JoinHook`` objects, then the ``device`` of the first is used. The
+        process group and device information is used for checking for non-
+        joined processes and for notifying processes to terminate if
+        ``throw_on_early_termination`` is eanbled, both of which using an all-
+        reduce.
+
+    Arguments:
+        join_hooks (List[_JoinHook]): a list of the :class:`_JoinHook` s to
+            use; the hooks are iterated over in the given order.
+
+        enable (bool): a flag enabling uneven input detection; setting to
+            ``False`` disables the context manager's functionality and should
+            only be set when the user knows the inputs will not be uneven
+            (default: ``True``).
+
+        throw_on_early_termination (bool): a flag controlling whether to raise
+            an exception upon detecting uneven inputs (default: ``False``).
+
+    """
+    def __init__(
+        self,
+        join_hooks: List[_JoinHook],
+        enable: bool = True,
+        throw_on_early_termination: bool = False,
+    ):
+        if len(join_hooks) == 0:
+            raise ValueError("The join context manager requires at least one join hook")
+        self._join_hooks = join_hooks
+        self._enable = enable
+        self._throw_on_early_termination = throw_on_early_termination
+        self._extract_dist_info()
+
+    def _extract_dist_info(self):
+        r"""
+        Extracts the process group and device information from the join hooks.
+
+        Preconditions:
+            ``self._join_hooks`` is not ``None`` and is non-empty.
+
+        Raises:
+            ValueError
+                If there are multiple conflicting ``process_group`` attributes
+                among the ``_JoinHook`` objects.
+
+        NOTE: The context manager uses the first specified device.
+        """
+        process_group = None
+        device = None
+        for join_hook in self._join_hooks:
+            if process_group is None:
+                process_group = join_hook.process_group
+            elif process_group != join_hook.process_group:
+                raise ValueError("Using join context manager with multiple process groups")
+            if device is None:
+                device = join_hook.device
```

Comment:
any reason we don't error out when devices do not match?

Reply:
I was thinking that having the same device is not a strict requirement like it is for the process group. If the process groups conflict, the all-reduces in the join implementation will not work as intended. If the devices conflict, that is fine as long as the chosen one (or even stronger, all of them) are valid devices accessible to the invoking process.

Currently, both DDP and ZeRO get their (default) device from the first parameter. However, for example in a model parallel case where the model parameters are spread across two devices (say process `0` uses `cuda:0` and `cuda:1`), even if DDP chose its (default) device to be `cuda:1` while ZeRO chose its (default) device to be `cuda:0`, there is no issue. IIUC, the join context manager can use either of the two for all-reduces with the other processes.